src/v/iceberg/conversion/values_json.cc

```diff
@@ -0,0 +1,266 @@
+        }
+    };
+
+    v = co_await std::visit(
+      field_decoder_visitor{
+        p,
+        field_map,
+      },
+      ft);
+}
+
+ss::future<> decode_struct(
+  experimental::serde::json::parser& p,
+  struct_value& sv,
+  const struct_type& st,
+  const json_conversion_ir::struct_field_map_t& field_map) {
+    if (p.token() != experimental::serde::json::token::start_object) {
+        throw value_conversion_exception(fmt::format(
+          "Expected start of JSON object for struct type but got {}",
+          p.token()));
+    }
+
+    // Placeholder fields.
+    for (size_t i = 0; i < st.fields.size(); ++i) {
+        sv.fields.emplace_back();
+    }
+
+    while (true) {
+        if (!co_await p.next()) {
+            throw value_conversion_exception(
+              "Failed to read next JSON token after start object");
+        }
+
+        if (p.token() == experimental::serde::json::token::end_object) {
+            for (size_t i = 0; i < st.fields.size(); ++i) {
+                if (st.fields[i]->required && !sv.fields[i].has_value()) {
+                    throw value_conversion_exception(fmt::format(
+                      "Required field {} is missing", st.fields[i]->name));
+                }
+            }
+            co_return;
+        }
+
+        if (p.token() != experimental::serde::json::token::key) {
+            throw value_conversion_exception(fmt::format(
+              "Expected key token in JSON object but got {}", p.token()));
+        }
+
+        auto key = p.value_string();
+
+        auto key_len = key.size_bytes();
+        iobuf_parser iop(std::move(key));
+        auto key_as_string = iop.read_string(key_len);
+
+        auto it = field_map.find(key_as_string);
+
+        if (it == field_map.end()) {
+            // Key not found in the mapping, skip it.
+            // todo throw if extra keys not allowed?
+            // skip the value
+            co_await p.next();
+            continue;
+        }
+
+        // Decode the value for this key.
+        if (!co_await p.next()) {
+            throw value_conversion_exception(
+              "Failed to read next JSON token after key");
+        }
+
+        co_await decode_field(
+          p,
+          sv.fields[it->second.field_pos],
+          st.fields[it->second.field_pos]->type,
+          it->second.nested_fields);
+    }
+}
+
+ss::future<> decode_list(
+  experimental::serde::json::parser& p,
+  list_value& lv,
+  const list_type& lt,
+  const json_conversion_ir::struct_field_map_t& field_map) {
+    if (p.token() != experimental::serde::json::token::start_array) {
+        throw value_conversion_exception(
+          "Expected start of JSON array for list type");
+    }
+
+    while (co_await p.next()) {
+        if (p.token() == experimental::serde::json::token::end_array) {
+            co_return;
+        }
+
+        co_await decode_field(
+          p, lv.elements.emplace_back(), lt.element_field->type, field_map);
+    }
```

Comment:
looks like a tight loop (similar in other places above i saw)? maybe it would benefit from using `ss::do_until` rather than the overhead of a coroutine? same in other places. could always benchmark later and see it pops up on a profile--but tight loops are more likely to be candidates. i think travis d had a p99 talk about this last year

Reply:
Will follow-up with a benchmark mirroring existing protobuf and avro and experiment with it.